### ISSUE TITLE
JavaParser consider as ClassOrInterfaceType 'java', 'io' in 'java.io.IOException'

### ISSUE THREAD

**Issue by corradomio** [javaparser/javaparser#2682]
Using JavaParser library 3.15.22, analyzing this code:

```java
package test;

class Test {
    private void ioexc() throws java.io.IOException {
        System.out.println("Hello World");
    }
    void test() {
        try {
            ioexc();
        }
        catch (java.io.IOException e) {
            e.printStackTrace();
        }
    }
}
```

using

```java
JavaParser jp = new JavaParser();

ParseResult<CompilationUnit> result = jp.parse(new File("D:\\Projects\\java.jext\\jext.main\\src\\test\\Test.java"));
Optional<CompilationUnit> optCu = result.getResult();

optCu.ifPresent(cu -> {
        cu.findAll(ClassOrInterfaceType.class)
            .forEach(cit -> {
                 System.out.println(cit.asString());
             });
});

```

I obtain the following list of 'ClassOrInterfaceType' objects:

```java
java.io.IOException
java.io
java
java.io.IOException
java.io
java
```

**Comment by matozoid** [comment]
The cause of this is that a parser does not know what the scope is. io might be a type, java might be a type. It guesses ClassOrInterfaceType. I think there is an issue somewhere to pick a new type of node for this.

The fix is probably to work on more granular node types, not ClassOrInterfaceType. To find out which ones, try examining the AST.

**Comment by corradomio** [comment]
I suppose it is necessary to handle the concept of ""Namespace/Package", because there are two cases to consider (at minimum):

```java
package test;

class Test {
     class Other {
     }
}
````

There are the following symbols to consider (when used in other parts of the code):

```java
test: a namespace
test.Test: a classOrInterface
test.Test.Other: another classOrInterface
```

However, in Java you can use OR the class name, resolved using the imports and "java.lang" OR the full qualified name of the class. A class as, for example 'Test.Other' can be FULL qualified class OR a subclass of 'Test' and 'Test' must be resolved using the standard rules.

I understand that this is a very complex problem to solve.

**Comment by matozoid** [comment]
Correct, and there is a tool for that: the symbol solver. It's best to take an hour or two to read through the docs if you haven't already: https://leanpub.com/javaparservisited and https://javaparser.org/inspecting-an-ast/

**Comment by corradomio** [comment]
I am sorry but I AM USING the symbol solver, and the problem is that the Symbol Solver is not able to solve namespaces. I am patching the SS to avoid these problems.

A "TypeSolver" is able to understand if a symbol is a "namespace":
1) in the source code, it is the "package"
2) in a jar file, a .class file has name "directory"/"classname".class or "directory"/"classname"$"innerclass".class. The "directory" is the namespace.

 The problem is here, I suppose: the typesolver has no concept of "namespace"

**Comment by MysterAitch** [comment]
After giving this some thought, I have come to the conclusion that our grammar needs rectifying _(personal opinion -- my rationale is shown below, but I'm more than open to being convinced otherwise!)_. 

Specifically, the parts to the left of the `ClassOrInterfaceType` should be a `Name`, not `ClassOrInterfaceType`. 

We then have `NameCategory.AMBIGUOUS_NAME` within the symbol solver to allow flagging that this name is potentially in need of disambiguation.

Thoughts from @matozoid and @ftomassetti would be appreciated here.

---

I suspect that the JLS grammar is where this has originated:

```
  https://docs.oracle.com/javase/specs/jls/se9/html/jls-4.html#jls-4.3

  ReferenceType:
    ClassOrInterfaceType
    TypeVariable
    ArrayType
  ClassOrInterfaceType:
    ClassType
    InterfaceType
  ClassType:
    {Annotation} Identifier [TypeArguments]
    ClassOrInterfaceType . {Annotation} Identifier [TypeArguments]
  InterfaceType:
    ClassType
  TypeVariable:
    {Annotation} Identifier
  ArrayType:
    PrimitiveType Dims
    ClassOrInterfaceType Dims
    TypeVariable Dims
  Dims:
    {Annotation} [ ] {{Annotation} [ ]}
```

In particular, I believe that `ClassOrInterfaceType . {Annotation} Identifier [TypeArguments]` refers to inner classes NOT the qualifying of COITs. This is where I spent quite a bit of time, until this epiphany occurred!



```java
    System.out.println("test");
//             ^^^^^^^ ClassOrInterfaceType
//  ^^^^^^^^^^ Qualification (Name)
//         ^^^ NameCategory.EXPRESSION_NAME (a field)
//  ^^^^^^     NameCategory.TYPE_NAME (ClassOrInterface name)

```

```java
    java.lang.System.out.println("test");
//                       ^^^^^^^ ClassOrInterfaceType
//  ^^^^^^^^^^^^^^^^^^^^ Qualification (Name)
//                   ^^^ NameCategory.EXPRESSION_NAME (a field)
//            ^^^^^^     NameCategory.TYPE_NAME (ClassOrInterface name)
//       ^^^^            NameCategory.PACKAGE_NAME
//  ^^^^                 NameCategory.PACKAGE_NAME
```

```java
    java.io.IOException
//          ^^^^^^^^^^^ ClassOrInterfaceType
//  ^^^^^^^ Qualification (Name)
//  ^^^^^^^ NameCategory.PACKAGE_NAME
```


```java
package com.example.test;
class X {
    class Y {
         class Z {
         }
    }
}

    com.example.test.X.Y.Z
//                       ^ ClassOrInterfaceType
//                     ^ ClassOrInterfaceType
//                   ^ ClassOrInterfaceType
//  ^^^^^^^^^^^^^^^^ Qualification (Name)
//  ^^^^^^^^^^^^^^^^ NameCategory.PACKAGE_NAME
//              ^^^^ NameCategory.PACKAGE_NAME
//      ^^^^^^^      NameCategory.PACKAGE_NAME
//  ^^^              NameCategory.PACKAGE_NAME

```

**Comment by matozoid** [comment]
Here's the issue: #2442 - not exactly the same, but the same kind of issue.

With epiphany, do you mean you found the possible nestings of qualifiers? I haven't gone into it that deep before. Do you have an approach for a fix?

**Comment by migliorabile** [comment]
I suppose it is possible to use the current context information.
IF we suppose that it is possible to compile the source file (without errors) each symbol is already solved using imports, current package "java.lang", inheritance e some other rule.
If it is true that it is necessary to follow exactly the resolution rules of Java, it is also true that if in import I have "java.util.List" and I am using the symbol "List" in a place where I must use a "class or interface type", this symbol can be ONLY "java.util.List". If there is the import "java.util.*", it is possible to try "java.util.List".

**Comment by migliorabile** [comment]
I suppose this is another instance of the same problem:

test class:

```java
package test;

class Test {
    private void ioexc() throws java.io.IOException {
        System.out.println("Hello World");
    }
}
```

parser & type solver:

```java
ParserConfiguration pc = new ParserConfiguration();
pc.setSymbolResolver(new JavaSymbolSolver(new ReflectionTypeSolver()));
JavaParser jp = new JavaParser(pc);
ParseResult<CompilationUnit> result = jp.parse(
    new File("D:\\Projects\\java.jext\\jext.main\\src\\test\\Test.java")
);
CompilationUnit cu = result.getResult().get();
```

Try to solve all 'NameExpr'

```java
cu.findAll(NameExpr.class)
            .forEach(ne -> {
                try {
                    System.out.println(ne.getName().asString());
                    ne.resolve();
                }
                catch (Exception e) {
                    System.out.println(e.getMessage());
                }
            });
```

It is not able to solve the symbol "System" of "System.out.println()"

**Comment by Aliaksei2022** [comment]
Still actual, any hope this will be fixed?

**Comment by ftomassetti** [comment]
> Still actual, any hope this will be fixed?

@Aliaksei2022 , in general, this is an open-source project, so feel free to make a contribution if there is a problem that you need solved.

In my opinion, from the point of view of the parser, it is impossible to distinguish between package names and types.

`a.b.c.d` could be:
* A type `d` in package `a.b.c`
* A type `d` annidated in class `c`, which would be in package `a.b`
* A type `d` anndated in class `c`, which is annidated in class `b`, which is in package `a`

We can solve that only resolving symbols, which is a futher process and it is not done at parsing time. So I agree with @MysterAitch that we could reflect this ambiguity in the name of the class used to represent these things
